test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/CreateMockitoMocksCallback.java

```diff
@@ -94,38 +71,52 @@ private Optional<Object> createMockAndSetTestField(Object testInstance, Field fi
         }
     }
 
-    static Object getBeanInstance(Object testInstance, Field field, Class<? extends Annotation> annotationType) {
+    /**
+     * Contextual reference of a normal scoped bean is a client proxy.
+     *
+     * @param testInstance
+     * @param field
+     * @param annotationType
+     * @return a contextual reference of a bean
+     */
+    static Object getContextualReference(Object testInstance, Field field, Class<? extends Annotation> annotationType) {
         Type fieldType = field.getGenericType();
-        Annotation[] qualifiers = getQualifiers(field);
         ArcContainer container = Arc.container();
         BeanManager beanManager = container.beanManager();
-        Set<Bean<?>> beans = beanManager.getBeans(fieldType, qualifiers);
-        if (beans.isEmpty()) {
+        Annotation[] qualifiers = getQualifiers(field, beanManager);
+
+        InstanceHandle<?> handle = container.instance(fieldType, qualifiers);
+        if (!handle.isAvailable()) {
             throw new IllegalStateException(
                     "Invalid use of " + annotationType.getTypeName() + " - could not resolve the bean of type: "
                             + fieldType.getTypeName() + ". Offending field is " + field.getName() + " of test class "
                             + testInstance.getClass());
         }
-        Bean<?> bean = beanManager.resolve(beans);
-        if (!beanManager.isNormalScope(bean.getScope())) {
+        if (!beanManager.isNormalScope(handle.getBean().getScope())) {
             throw new IllegalStateException(
                     "Invalid use of " + annotationType.getTypeName()
-                            + " - the injected bean does not declare a CDI normal scope but: " + bean.getScope().getName()
+                            + " - the injected bean does not declare a CDI normal scope but: "
+                            + handle.getBean().getScope().getName()
                             + ". Offending field is " + field.getName() + " of test class "
                             + testInstance.getClass());
         }
-        return container.instance((InjectableBean<?>) bean).get();
+        return handle.get();
+    }
+
+    static Class<?> getImplementationClass(Object contextualReference) {
+        // Unwrap the client proxy if needed
+        Object contextualInstance = ClientProxy.unwrap(contextualReference);
+        // If the contextual instance is an intercepted subclass then mock the extended implementation class
+        return contextualInstance instanceof Subclass ? contextualInstance.getClass().getSuperclass()
```

Comment:
This means that for a producer method that returns an interface the mock is created from the implementation class.

```
@Produces
List<String> produce() {
  new ArrayList<>(); // -> mock is created from ArrayList
}
```

Reply:
If you decide to make any more changes, you could probably improve the commit message somewhat, because it says "return type of a producer method" and here, you show that that's imprecise. It's a "class of a producer method return value" (which is a mouthful, yes...).